docs/designs/topology.md

```diff
@@ -0,0 +1,916 @@
+# Topology-Aware Scheduling - Grove Operator Design
+
+## Overview
+
+This document defines the design for implementing topology-aware scheduling in the Grove operator.
+
+**Motivation**: Topology-aware scheduling is critical for Grove's multinode inference workloads because these
+applications require:
+
+- **Network Locality**: Proximity improves high-bandwidth communication between leaders and their respective workers (
+  prefill and decode, etc)
+    - **Coordinated Placement**: Related components (e.g., model shards) perform better when co-located within the same
+      topology domain
+  - **Latency Optimization**: Minimizing network hops between interdependent inference components improves end-to-ends
+      performance
+
+**Design Approach**: This design introduces a flexible topology system with three main components:
+
+1. **TopologyDomain CRD**: Admin-configured cluster topology hierarchy mapping friendly names to node labels
+    2. **Operator Configuration**: Selects active topology via `--topology-domain-name` argument
+    3. **TopologyConstraint**: User-specified packing requirements in workloads (PodCliqueSet, PodCliqueScalingGroup,
+       PodClique)
+
+**Key Feature**: Grove provides automatic out-of-box topology optimization by generating preferred packing constraints
+at all levels, even without user configuration. Users can optionally specify required constraints for strict placement
+requirements.
+
+## Goals
+
+- Provide flexible, cluster-agnostic topology hierarchy definition via TopologyDomain CRD
+    - Enable packing constraints for network locality across all Grove scalable resources
+    - Support multiple topology configurations for different environments
+    - Automatic Kueue Topology generation for KAI scheduler integration
+    - Immutable topology configuration ensuring scheduling consistency
+    - Hierarchical constraint validation (child stricter than parent)
+
+## Non-Goals
+
+- Spread constraints across topology domains (ReplicaSpreadDomain)
+    - Root domain constraints for entire resource (RootDomain)
+    - Ratio-based affinity groups between scaling groups (AffinityGroups with PackRatio)
+    - Dynamic topology reconfiguration after creation
+    - Per-workload topology domain selection
+    - Automatic topology inference from workload characteristics
+
+## Proposal
+
+### High-Level Approach
+
+Grove implements topology-aware scheduling through three key components:
+
+1. **TopologyDomain CRD**: Cluster-scoped resource defining topology hierarchy
+    - Admin creates TopologyDomain with ordered list of topology levels
+    - Each level maps friendly name (e.g., "rack") to node label key (e.g., "topology.kubernetes.io/rack")
+    - Multiple TopologyDomains supported for different environments
+
+    2. **Operator Configuration**: References TopologyDomain by name
+        - Operator argument `--topology-domain-name=default` selects which TopologyDomain to use
```

Comment:
Another option is to have a annotation in the `TopologyDomain` CR "is-default: true" that only one topology has, and grove knows which one is the default.

A benefit of that, is that you can change the used topology without restarting the grove controller.
The bad side is that you need to handle what happen if there are 2 instances or more with that annotation.

Thats how default StorageClass is implemented in K8S: https://kubernetes.io/docs/tasks/administer-cluster/change-default-storage-class/